README.md
loresim
=======

Long read simulation.

Source
------

The loresim source code is hosted on github:

	git@github.com:gt1/loresim.git

Compilation of loresim
----------------------

loresim needs libmaus2 [https://github.com/gt1/libmaus2] . When libmaus2
is installed in ${LIBMAUSPREFIX} then biobambam2 can be compiled and
installed in ${HOME}/biobambam2 using

	- autoreconf -i -f
	- ./configure --with-libmaus2=${LIBMAUSPREFIX} \
		--prefix=${HOME}/biobambam2
	- make install

Contained programs
------------------

*genseq*: generates a sequence of random DNA blocks in FastA format. Blocks are repeated with a given probability. The program has the following options

- blocksizeavg: average block size (default: 8k)
- blocksizestd: block size standard deviation (default: 4k)
- blockmin: minimum block size (default: 512)
- repfrag: block repeat probability (default: 0.01)
- reprepfrag: probability that repeated block is chosen from list of already repeated blocks (default: 0.5)
- numblocks: number of blocks to be produced (default: 1k)
- erate: rate of error spiked into output blocks (default: 0.001)
- eratesq: error rate standard deviation (default: 0.001)
- substrate: fraction of substitutions in introduced base errors
- delrate: fraction of deletions in introduced base errors
- insrate: fraction of insertions in introduced base errors

*fastareformat*: reformats FastA or FastQ so it becomes valid input for the DAZZ_DB tool fasta2DB. This includes changing read names and wrapping long lines.

*loresim*: generates a sequence of random DNA reads from a given sequence. For this the input sequence is traversed a number of times and split into pieces which are subsequently either
modified within a certain error range or dropped with a certain rate. The program can switch between a low and high error profile during read generation. Reads are randomly taken from the forward or reverse strand.
The program has the following options

- readlenavg: average read length (default: 15k)
- readlenstddev: standard deviation for read length (default: 3k)
- droprate: probability for dropping a read (default: 0.01)
- numtraversals: number of traversals of the input sequence (default: 1)
- startlowprob: probability of starting a read in low error mode (default: 0.7)
- keeplowstate: probability of staying in low error mode when in low error mode (default: 0.9998)
- keephighstate: probability of staying in high error mode when in high error mode (default: 0.995)
- eratelow: low error mode error rate average
- eratelowstddev: standard deviation of error rate in low error mode
- eratehigh: high error mode error rate average
- eratehighstddev: standard deviation of error rate in high error mode


ChangeLog
loresim (0.0.2-1) unstable; urgency=medium

  * allow fastq input in fastareformat 

 -- German Tischler <gt1@sanger.ac.uk>  Wed, 17 Jun 2015 14:36:21 +0200

loresim (0.0.1-1) unstable; urgency=medium

  * add documentation 

 -- German Tischler <gt1@sanger.ac.uk>  Mon, 08 Jun 2015 16:11:56 +0200

loresim (0.0.0-1) UNRELEASED; urgency=medium

  * Initial release

 -- German Tischler <tischler@mpi-cbg.de>  Mon, 08 Jun 2015 16:10:29 +0200


simulate.sh
#! /bin/bash
if [ $# -lt 3 ] ; then
	echo "usage: $0 sequence_in.fasta reads_out.fasta #traversals"
	exit 1
fi

if [ ! -f src/loresim ] ; then
	echo "Program src/loresim does not exist. Please compile it first."
	exit 1
fi

IN=$1

if [ ! -f $IN ] ; then
	if [ ! -f src/genseq ] ; then
		echo "Program src/genseq does not exist. Please compile it first."
		exit 1
	fi
	
	src/genseq repfrag=0.1 numblocks=64k > $IN
fi

OUT=$2
TRAVERSALS=$3

OUTSHORT=${OUT%.fasta}
OUTSHORT=${OUTSHORT%.fa}
OUTSHORT=${OUTSHORT}_short_names.fasta

src/loresim numtraversals=$TRAVERSALS <$IN >$OUT
awk < $OUT '/^>/ { print $1 " " $2 } !/^>/ { print }' >${OUTSHORT}

if [ -f ../DAZZ_DB/fasta2DB ] ; then
	if [ -f ${OUTSHORT%.fasta}.db ] ; then
		../DAZZ_DB/DBrm ${OUTSHORT%.fasta}.db
	fi
	../DAZZ_DB/fasta2DB ${OUTSHORT%.fasta}.db ${OUTSHORT}

	if [ -f ../DALIGNER/HPCdaligner ] ; then
		rm -f ${OUTSHORT%.fasta}*.las
		../DALIGNER/HPCdaligner ${OUTSHORT%.fasta}.db | PATH=$PATH:../DALIGNER bash

		rm -f ${OUTSHORT%.fasta}.${OUTSHORT%.fasta}.C*.las
		rm -f ${OUTSHORT%.fasta}.${OUTSHORT%.fasta}.N*.las
		
		if [ -f ../../celamy2/LAcelamy ] ; then
			../../celamy2/LAcelamy ${OUTSHORT%.fasta}.db ${OUTSHORT%.fasta}.db ${OUTSHORT%.fasta}.1.las
		fi
	fi
fi


bumpversion.sh
#! /bin/bash
VERSION=`grep AC_INIT < configure.ac | awk -F',' '{print $2}'`
FIRST=`echo $VERSION | awk -F'.' '{print $1}'`
SECOND=`echo $VERSION | awk -F'.' '{print $2}'`
THIRD=`echo $VERSION | awk -F'.' '{print $3}'`
NEXTTHIRD=`expr ${THIRD} + 1`
export DEBEMAIL=gt1@sanger.ac.uk
export DEBFULLNAME="German Tischler"

function cleanup
{
	if [ ! -z "${COMMITFILE}" ] ; then
		if [ -f "${COMMITFILE}" ] ; then
			rm -f "${COMMITFILE}"
		fi
	fi
}

COMMITFILE=commit_msg_$$.txt

trap cleanup EXIT SIGINT SIGTERM

# make sure we have the latest version
git pull

# create commit log message
joe "${COMMITFILE}"

if [ ! -s "${COMMITFILE}" ] ; then
	echo "Empty commit log, aborting"
	exit 1
fi

# update to next minor version
awk -v first=${FIRST} -v second=${SECOND} -v third=${THIRD} '/^AC_INIT/ {gsub(first"."second"."third,first"."second"."third+1);print} ; !/^AC_INIT/{print}' < configure.ac > configure.ac.tmp
mv configure.ac.tmp configure.ac

# update change log
CHANGELOG=ChangeLog dch --distribution unstable -v ${FIRST}.${SECOND}.${NEXTTHIRD}-1

# commit files
git add configure.ac ChangeLog

git commit -F "${COMMITFILE}"
git push

TAG=loresim_${FIRST}_${SECOND}_${NEXTTHIRD}
git tag -a ${TAG} -m "loresim version ${FIRST}_${SECOND}_${NEXTTHIRD}"
git push origin ${TAG}

git checkout master
VERSION=`grep <configure.ac "AC_INIT" | perl -p -e "s/.*AC_INIT\(//" | awk -F ',' '{print $2}'`
DATE=`date +"%Y%m%d%H%M%S"`
RELEASE=${VERSION}-release-${DATE}
git checkout -b ${RELEASE}-branch master
PATH=/software/hpag/autotools/bin:${PATH} autoreconf -i -f
ADDFILES="INSTALL Makefile.in aclocal.m4 autom4te.cache compile config.guess config.h.in config.sub configure depcomp install-sh ltmain.sh m4/libtool.m4 m4/ltoptions.m4 m4/ltsugar.m4 m4/ltversion.m4 m4/lt~obsolete.m4 missing src/Makefile.in"
mv .gitignore .gitignore_
git add ${ADDFILES}
git commit -m "Release ${RELEASE}"
mv .gitignore_ .gitignore
git tag ${RELEASE}
git push origin ${RELEASE}
git checkout master
git branch -D ${RELEASE}-branch

exit 0


src/genseq.cpp
/*
    loresim
    Copyright (C) 2015 German Tischler

    This program is free software: you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation, either version 3 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program.  If not, see <http://www.gnu.org/licenses/>.
*/
#include <libmaus2/util/ArgInfo.hpp>
#include <libmaus2/random/GaussianRandom.hpp>
#include <libmaus2/fastx/acgtnMap.hpp>
#include <set>

enum err_type_enum
{
    err_subst,
    err_ins,
    err_del
};

static bool contains(std::vector<err_type_enum> const & V, err_type_enum type)
{
    for ( uint64_t i = 0; i < V.size(); ++i )
        if ( V[i] == type )
            return true;
    return false;
}

enum state_enum
{
    state_error_low = 0,
    state_error_high = 1
};


int main(int argc, char * argv[])
{
	try
	{
		libmaus2::util::ArgInfo const arginfo(argc,argv);
		
		uint64_t const blocksizeavg = arginfo.getValueUnsignedNumeric<uint64_t>("blocksizeavg",8*1024);
		uint64_t const blocksizestd = arginfo.getValueUnsignedNumeric<uint64_t>("blocksizestd",4*1024);
		uint64_t const blockmin = arginfo.getValueUnsignedNumeric<uint64_t>("blockmin",512);
		double repfrag = arginfo.getValue<double>("repfrag",0.01);	
		double reprepfrag = arginfo.getValue<double>("reprepfrag",0.5);
		uint64_t const numblocks = arginfo.getValueUnsignedNumeric<uint64_t>("numblocks",1024);
		double const eratesq = arginfo.getValue<double>("eratesq",0.001);
		double const erateavg = arginfo.getValue<double>("erateavg",0.001);
		double substrate = arginfo.getValue<double>("substrate",.2);
		double delrate = arginfo.getValue<double>("delrate",.3);
		double insrate = arginfo.getValue<double>("insrate",.5);
		
		std::vector<uint64_t> blockids;
		std::vector<std::string> blocks;
		std::set<uint64_t> repblockset;
		libmaus2::random::Random::setup();
		
		for ( uint64_t i = 0; i < numblocks; ++i )
		{
			if ( blocks.size() && libmaus2::random::UniformUnitRandom::uniformUnitRandom() < repfrag )
			{
				uint64_t repblock;
				
				if ( repblockset.size() && libmaus2::random::UniformUnitRandom::uniformUnitRandom() < reprepfrag )
				{
					uint64_t repoff = libmaus2::random::Random::rand64() % repblockset.size();
					std::set<uint64_t>::const_iterator ita = repblockset.begin();
					while ( repoff-- )
						++ita;
					repblock = *ita;
				}
				else
				{
					repblock = libmaus2::random::Random::rand64() % blocks.size();
				}

				std::cerr << "[D] rep " << repblock << " of size " << blocks[repblock].size() << std::endl;
				blockids.push_back(repblock);				
				repblockset.insert(repblock);
			}
			else
			{
				double dlength = -1;

				while ( dlength <= blockmin )
					dlength = libmaus2::random::GaussianRandom::random(blocksizestd,blocksizeavg);

				uint64_t const length = dlength;
				
				std::vector<char> V(length);
				for ( uint64_t j = 0; j < V.size(); ++j )
					V[j] = libmaus2::fastx::remapChar(libmaus2::random::Random::rand8()%4);
					
				std::cerr << "[D] new block of size " << V.size() << std::endl;
				
				blockids.push_back(blocks.size());
				blocks.push_back(std::string(V.begin(),V.end()));
			}
		}
		
		std::cerr << "[D] all blocks generated" << std::endl;
		
		for ( uint64_t i = 0; i < blockids.size(); ++i )
		{
			
			std::string sub = blocks[blockids[i]];
			
			std::map<uint64_t,std::vector<err_type_enum> > errM;
			for ( uint64_t i = 0; i < sub.size(); ++i )
			    errM[i] = std::vector<err_type_enum>(0);

			std::map<int,double> err_prob_cumul;
			double ratesum = substrate + delrate + insrate;
			substrate /= ratesum;
			delrate /= ratesum;
			insrate /= ratesum;
        
			err_prob_cumul[err_subst] = substrate;
			err_prob_cumul[err_del] = substrate + delrate;
			err_prob_cumul[err_ins] = substrate + delrate + insrate;

			double erate = -1;
			while ( (erate = libmaus2::random::GaussianRandom::random(eratesq,erateavg)) < 0 )
			{    
			}
			    
			uint64_t const numerr = std::floor(erate * sub.size() + 0.5);

			uint64_t errplaced = 0;
			    
			while ( errplaced < numerr )
			{
				double const p = libmaus2::random::UniformUnitRandom::uniformUnitRandom();
				uint64_t const errpos = std::floor(libmaus2::random::UniformUnitRandom::uniformUnitRandom() * (sub.size()-1) + 0.5);
				
				if ( p < err_prob_cumul[err_subst]  )
				{
					// no error yet?
					if ( errM.find(errpos) == errM.end() )
					{
						errM[errpos].push_back(err_subst);
						errplaced++;
					}
					// no error so far a deletion or substitution?
					else if ( 
						!contains(errM.find(errpos)->second,err_del)
						&&
						!contains(errM.find(errpos)->second,err_subst)
					)
					{
						errM[errpos].push_back(err_subst);
						errplaced++;                        
					}
				}
				else if ( p < err_prob_cumul[err_del] )
				{
					// no error yet?
					if ( errM.find(errpos) == errM.end() )
					{
						errM[errpos].push_back(err_del);
						errplaced++;
					}                
					// no error so far a deletion or substitution?
					else if ( 
						!contains(errM.find(errpos)->second,err_del)
						&&
						!contains(errM.find(errpos)->second,err_subst)
					)
					{
						errM[errpos].push_back(err_subst);
						errplaced++;                        
					}
				}
				else
				{
					errM[errpos].push_back(err_ins);
					errplaced++;
				}
			}

			std::ostringstream baseostr;
			for ( std::map<uint64_t,std::vector<err_type_enum> >::const_iterator ita = errM.begin(); ita != errM.end(); ++ita )
			{
				uint64_t const pos = ita->first;
				std::vector<err_type_enum> const & errV = ita->second;
                    
				// insertions                
				for ( uint64_t i = 0; i < errV.size(); ++i )
					if ( errV[i] == err_ins )
					{
						int const v = libmaus2::random::Random::rand8() & 3;
						int insbase = -1;
						switch ( v )
						{
							case 0: insbase = 'A'; break;
							case 1: insbase = 'C'; break;
							case 2: insbase = 'G'; break;
							case 3: insbase = 'T'; break;
						}
						baseostr.put(insbase);
					}

					// base not deleted?
					if ( !contains(errV,err_del) )
					{
						if ( contains(errV,err_subst) )
						{
							int const v = libmaus2::random::Random::rand8() & 3;
							int insbase = -1;
							switch ( v )
							{
								case 0: insbase = 'A'; break;
								case 1: insbase = 'C'; break;
								case 2: insbase = 'G'; break;
								case 3: insbase = 'T'; break;
							}                    
							baseostr.put(insbase);
						}
						else
						{
							baseostr.put(sub[pos]); 
						}
					}
					else
					{
					}
			}
			    
			std::cout << ">seq_" << blockids[i] << "_" << baseostr.str().size() << "\n";
			std::cout << baseostr.str() << "\n";	
		}
	}
	catch(std::exception const & ex)
	{
		std::cerr << ex.what() << std::endl;
		throw ex;
	}
}


src/fastareformat.cpp
/*
    loresim
    Copyright (C) 2015 German Tischler

    This program is free software: you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation, either version 3 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program.  If not, see <http://www.gnu.org/licenses/>.
*/
#include <libmaus2/fastx/StreamFastAReader.hpp>
#include <libmaus2/fastx/StreamFastQReader.hpp>
#include <libmaus2/aio/PosixFdInputStream.hpp>
#include <libmaus2/fastx/SpaceTable.hpp>

template<typename reader_type>
void fastareformat(libmaus2::util::ArgInfo const & arginfo, reader_type & reader)
{
	typename reader_type::pattern_type pattern;
	
	std::string const prolog = arginfo.getUnparsedValue("prolog","L");
	uint64_t readid = arginfo.getValueUnsignedNumeric<uint64_t>("readidbase",0);
	uint64_t const cols = arginfo.getValueUnsignedNumeric<uint64_t>("cols",80);
	libmaus2::fastx::SpaceTable const ST;
	
	while ( reader.getNextPatternUnlocked(pattern) )
	{
		std::string const s = pattern.spattern;
		char const * c = s.c_str();
				
		std::cout << '>' << prolog << '/' << (readid++) << '/' << 0 << '_' << s.size() << " RQ=0.851 " << pattern.sid << "\n";
		uint64_t low = 0;
		
		while ( low < s.size() )
		{
			uint64_t const high = std::min(low+cols,static_cast<uint64_t>(s.size()));
			
			std::cout.write(c+low,high-low);
			std::cout.put('\n');
			
			low = high;
		}
	}
}

int main(int argc, char * argv[])
{
	try
	{
		libmaus2::util::ArgInfo const arginfo(argc,argv);

		libmaus2::aio::PosixFdInputStream PFIS(STDIN_FILENO,64*1024,1024);
		
		int const c = PFIS.peek();
		
		if ( c == '>' || c == std::istream::traits_type::eof() )
		{
			libmaus2::fastx::StreamFastAReaderWrapper in(PFIS);	
			fastareformat(arginfo,in);	
		}
		else if ( c == '@' )
		{
			libmaus2::fastx::StreamFastQReaderWrapper in(PFIS);	
			fastareformat(arginfo,in);			
		}
		else
		{
			libmaus2::exception::LibMausException lme;
			lme.getStream() << "Unknown input file format, first character " << static_cast<char>(c) << std::endl;
			lme.finish();
			throw lme;
		}	
	}
	catch(std::exception const & ex)
	{
		std::cerr << ex.what() << std::endl;
		return EXIT_FAILURE;
	}
}


src/loresim.cpp
/*
    loresim
    Copyright (C) 2015 German Tischler

    This program is free software: you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation, either version 3 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program.  If not, see <http://www.gnu.org/licenses/>.
*/
#include <libmaus2/util/ArgInfo.hpp>
#include <libmaus2/random/GaussianRandom.hpp>
#include <libmaus2/fastx/StreamFastAReader.hpp>

enum err_type_enum
{
    err_subst,
    err_ins,
    err_del
};

static bool contains(std::vector<err_type_enum> const & V, err_type_enum type)
{
    for ( uint64_t i = 0; i < V.size(); ++i )
        if ( V[i] == type )
            return true;
    return false;
}

enum state_enum
{
    state_error_low = 0,
    state_error_high = 1
};

int main(int argc, char ** argv)
{
    try
    {
        libmaus2::util::ArgInfo const arginfo(argc,argv);
        uint64_t randseed = arginfo.getValueUnsignedNumeric<uint64_t>("randomseed",time(0));
        libmaus2::random::Random::setup(randseed);

        std::map<int,double> err_prob_cumul;
        
        double substrate = arginfo.getValue<double>("substrate",.2);
        double delrate = arginfo.getValue<double>("delrate",.3);
        double insrate = arginfo.getValue<double>("insrate",.5);
        double ratesum = substrate + delrate + insrate;
        substrate /= ratesum;
        delrate /= ratesum;
        insrate /= ratesum;
        
        err_prob_cumul[err_subst] = substrate;
        err_prob_cumul[err_del] = substrate + delrate;
        err_prob_cumul[err_ins] = substrate + delrate + insrate;
        
        if ( std::abs ( err_prob_cumul.find(err_ins)->second - 1.0 ) > 1e-3 )
        {
            libmaus2::exception::LibMausException lme;
            lme.getStream() << "subst, del and insrate cannot be normalised to sum 1" << std::endl;
            lme.finish();
            throw lme;
        }

        double const droprate = arginfo.getValue<double>("droprate",0.01);        
        uint64_t const numtraversals = arginfo.getValueUnsignedNumeric<uint64_t>("numtraversals",1);
        double const readlenavg = arginfo.getValue<double>("readlenavg", 15000);
        double const readlenstddev = arginfo.getValue<double>("readlenstddev", 3000);
        double const keeplowstate = arginfo.getValue<double>("keeplowstate", 0.9998);
        double const keephighstate = arginfo.getValue<double>("keeplowstate", 0.995);
        double const startlowprob = arginfo.getValue<double>("startlowprob", 0.7);

        std::map<int,double> state_erate;
        std::map<int,double> state_erate_sq;
       
        state_erate[state_error_low] = arginfo.getValue<double>("eratelow", 0.15);
        state_erate[state_error_high] = arginfo.getValue<double>("eratehigh",0.25);
        state_erate_sq[state_error_low] = arginfo.getValue<double>("eratelowstddev", 0.03);
        state_erate_sq[state_error_high] = arginfo.getValue<double>("eratehighstddev", 0.04);
    
        libmaus2::fastx::StreamFastAReaderWrapper SFAR(std::cin);
        libmaus2::fastx::StreamFastAReaderWrapper::pattern_type pattern;
        std::ostringstream seqstr;
        
        while ( SFAR.getNextPatternUnlocked(pattern) )
            seqstr << pattern.spattern;
        
        std::string const seq = seqstr.str();
        
        std::map< int, std::map<int,double> > state_change_map;
        std::map< int, double > state_start_map;
        
        state_change_map[state_error_low][state_error_low ] = keeplowstate;
        state_change_map[state_error_low][state_error_high] = 1-state_change_map[state_error_low][state_error_low ];
        state_change_map[state_error_high][state_error_high] = keephighstate;
        state_change_map[state_error_high][state_error_low] = 1-state_change_map[state_error_high][state_error_high ];
        state_start_map[state_error_low] = startlowprob;
        state_start_map[state_error_high] = 1 - state_start_map[state_error_low];
        
        uint64_t runid = 0;
        uint64_t readid = 0;

        for ( uint64_t trav = 0; trav < numtraversals; ++trav )
        {
            uint64_t pp = 0;
            typedef std::pair<uint64_t,uint64_t> upair;
            std::vector<upair> poslenvec;
            
            bool const strand = libmaus2::random::Random::rand8() & 1;
            
            while ( pp < seq.size() )
            {
                int64_t len = libmaus2::random::GaussianRandom::random(readlenstddev,readlenavg);
                
                len = std::min(len,static_cast<int64_t>(seq.size()-pp));
                
                poslenvec.push_back(upair(pp,len));
                
                pp += len;
            }
        
            for ( uint64_t z = 0; z < poslenvec.size(); ++z )
            {
                if ( libmaus2::random::UniformUnitRandom::uniformUnitRandom() < droprate )
                    continue;
            
                std::ostringstream baseostr;
                std::ostringstream errostr;
                
                uint64_t const len = poslenvec[z].second;
                uint64_t const pos = strand ? poslenvec[z].first : (seq.size() - poslenvec[z].first - len);

                std::string sub = strand ? seq.substr(pos,len) : libmaus2::fastx::reverseComplementUnmapped(seq.substr(pos,len));
                
                std::vector < state_enum > states;
                state_enum state = (libmaus2::random::UniformUnitRandom::uniformUnitRandom() < state_start_map[state_error_low]) ? state_error_low : state_error_high;
                for ( uint64_t i = 0; i < sub.size(); ++i )
                {
                    states.push_back(state);
                    
                    if ( libmaus2::random::UniformUnitRandom::uniformUnitRandom() < state_change_map[state][state_error_low] )
                        state = state_error_low;
                    else
                        state = state_error_high;
                }
                
                std::map<uint64_t,std::vector<err_type_enum> > errM;
                for ( uint64_t i = 0; i < sub.size(); ++i )
                    errM[i] = std::vector<err_type_enum>(0);
                
                
                uint64_t low = 0;
                while ( low < sub.size() )
                {
                    uint64_t high = low;
                    while ( high < sub.size() && states[high] == states[low] )
                        ++high;

                    double erate = -1;
                    while ( (erate = libmaus2::random::GaussianRandom::random(state_erate_sq[states[low]],state_erate[states[low]])) < 0 )
                    {
                    
                    }
                    
                    uint64_t const numerr = std::floor(erate * (high-low) + 0.5);
                        
                    errostr << "einterval(" << "[" << low << "," << high << ")" << ",erate=" << erate << ",numerr=" << numerr << ')' << ';';
                    
                    uint64_t errplaced = 0;
                    
                    while ( errplaced < numerr )
                    {
                        double const p = libmaus2::random::UniformUnitRandom::uniformUnitRandom();
                        uint64_t const errpos = std::floor(libmaus2::random::UniformUnitRandom::uniformUnitRandom() * (high-low-1) + 0.5) + low;
                        
                        if ( p < err_prob_cumul[err_subst]  )
                        {
                            // no error yet?
                            if ( errM.find(errpos) == errM.end() )
                            {
                                errM[errpos].push_back(err_subst);
                                errplaced++;
                            }
                            // no error so far a deletion or substitution?
                            else if ( 
                                !contains(errM.find(errpos)->second,err_del)
                                &&
                                !contains(errM.find(errpos)->second,err_subst)
                            )
                            {
                                errM[errpos].push_back(err_subst);
                                errplaced++;                        
                            }
                        }
                        else if ( p < err_prob_cumul[err_del] )
                        {
                            // no error yet?
                            if ( errM.find(errpos) == errM.end() )
                            {
                                errM[errpos].push_back(err_del);
                                errplaced++;
                            }                
                            // no error so far a deletion or substitution?
                            else if ( 
                                !contains(errM.find(errpos)->second,err_del)
                                &&
                                !contains(errM.find(errpos)->second,err_subst)
                            )
                            {
                                errM[errpos].push_back(err_subst);
                                errplaced++;                        
                            }
                        }
                        else
                        {
                            errM[errpos].push_back(err_ins);
                            errplaced++;
                        }
                    }
                    
                    low = high;
                }
             
                errostr << "pos=" << pos << ';';
                errostr << "strand=" << strand << ';';
                
                for ( std::map<uint64_t,std::vector<err_type_enum> >::const_iterator ita = errM.begin(); ita != errM.end(); ++ita )
                {
                    uint64_t const pos = ita->first;
                    std::vector<err_type_enum> const & errV = ita->second;
                    
                    // insertions                
                    for ( uint64_t i = 0; i < errV.size(); ++i )
                        if ( errV[i] == err_ins )
                        {
                            int const v = libmaus2::random::Random::rand8() & 3;
                            int insbase = -1;
                            switch ( v )
                            {
                                case 0: insbase = 'A'; break;
                                case 1: insbase = 'C'; break;
                                case 2: insbase = 'G'; break;
                                case 3: insbase = 'T'; break;
                            }
                            baseostr.put(insbase);
                            errostr << 'i' << static_cast<char>(insbase);
                        }

                    // base not deleted?
                    if ( !contains(errV,err_del) )
                    {
                        if ( contains(errV,err_subst) )
                        {
                            int const v = libmaus2::random::Random::rand8() & 3;
                            int insbase = -1;
                            switch ( v )
                            {
                                case 0: insbase = 'A'; break;
                                case 1: insbase = 'C'; break;
                                case 2: insbase = 'G'; break;
                                case 3: insbase = 'T'; break;
                            }                    
                            baseostr.put(insbase);
                            errostr << 's' << static_cast<char>(insbase);
                        }
                        else
                        {
                           baseostr.put(sub[pos]); 
                           errostr << 'o';
                        }
                    }
                    else
                    {
                        errostr.put('d');
                    }
                }
                
                std::cout << '>' << 'L' << runid << '/' << (readid++) << '/' << 0 << '_' << baseostr.str().size() << " RQ=0.851 " << errostr.str() << '\n';

                uint64_t b_low = 0;
                std::string const bases = baseostr.str();
                
                while ( b_low < bases.size() )
                {
                    uint64_t const high = std::min(b_low + 80, static_cast<uint64_t>(bases.size()));
                    
                    std::cout.write(bases.c_str() + b_low, high-b_low);
                    std::cout.put('\n');
                
                    b_low = high;
                }
            }
        }
        
    }
    catch(std::exception const & ex)
    {
        std::cerr << ex.what() << std::endl;
        return EXIT_FAILURE;
    }
}
